File: zig/src/common/entity.zig
const std = @import("std");
const string_mod = @import("string.zig");

pub const EntityFreq = struct {
    name: []const u8,
    frequency: usize,
    entity_type: EntityType,
};

pub const EntityType = enum {
    person,
    project,
    location,
    uncertain,
};

pub const ENTITY_STOPLIST = std.StaticStringMap(void).initComptime(.{
    .{ "The", {} },      .{ "This", {} },     .{ "That", {} },      .{ "When", {} },
    .{ "Where", {} },    .{ "What", {} },     .{ "Why", {} },       .{ "Who", {} },
    .{ "Which", {} },    .{ "How", {} },      .{ "Then", {} },      .{ "There", {} },
    .{ "Here", {} },     .{ "Now", {} },      .{ "Just", {} },      .{ "Also", {} },
    .{ "Some", {} },     .{ "Such", {} },     .{ "Each", {} },      .{ "Every", {} },
    .{ "Monday", {} },   .{ "Tuesday", {} },  .{ "Wednesday", {} }, .{ "Thursday", {} },
    .{ "Friday", {} },   .{ "Saturday", {} }, .{ "Sunday", {} },    .{ "January", {} },
    .{ "February", {} }, .{ "March", {} },    .{ "April", {} },     .{ "May", {} },
    .{ "June", {} },     .{ "July", {} },     .{ "August", {} },    .{ "September", {} },
    .{ "October", {} },  .{ "November", {} }, .{ "December", {} },
});

pub fn candidateEntityWords(text: []const u8, allocator: std.mem.Allocator) ![]const []const u8 {
    var candidates: std.ArrayList([]const u8) = .empty;
    errdefer {
        for (candidates.items) |item| allocator.free(item);
        candidates.deinit(allocator);
    }

    var word_start: ?usize = null;
    var i: usize = 0;

    while (i <= text.len) : (i += 1) {
        const ch: u8 = if (i < text.len) text[i] else ' ';
        if (isEntityWordChar(ch)) {
            if (word_start == null) word_start = i;
        } else {
            if (word_start) |start| {
                const word = text[start..i];
                if (word.len >= 2 and isCapitalized(word)) {
                    if (!ENTITY_STOPLIST.has(word)) {
                        const duped = try allocator.dupe(u8, word);
                        try candidates.append(allocator, duped);
                    }
                }
            }
            word_start = null;
        }
    }

    return candidates.toOwnedSlice(allocator);
}

fn isEntityWordChar(c: u8) bool {
    return std.ascii.isAlphanumeric(c) or c == '_' or c == '-' or c == '\'' or c == '.';
}

fn isCapitalized(word: []const u8) bool {
    if (word.len == 0) return false;
    return std.ascii.isUpper(word[0]);
}

pub fn extractEntities(content: []const u8, allocator: std.mem.Allocator, min_frequency: usize) ![]EntityFreq {
    var freq_map = std.StringHashMap(usize).init(allocator);
    defer {
        var it = freq_map.iterator();
        while (it.next()) |entry| allocator.free(entry.key_ptr.*);
        freq_map.deinit();
    }

    const candidates = try candidateEntityWords(content, allocator);
    defer {
        for (candidates) |c| allocator.free(c);
        allocator.free(candidates);
    }

    for (candidates) |entity| {
        const key = try allocator.dupe(u8, entity);
        const result = try freq_map.getOrPut(key);
        if (result.found_existing) {
            allocator.free(key);
        }
        result.value_ptr.* += 1;
    }

    var results: std.ArrayList(EntityFreq) = .empty;
    errdefer results.deinit(allocator);

    var iter = freq_map.iterator();
    while (iter.next()) |entry| {
        if (entry.value_ptr.* >= min_frequency) {
            try results.append(allocator, .{
                .name = entry.key_ptr.*,
                .frequency = entry.value_ptr.*,
                .entity_type = .uncertain,
            });
        }
    }

    std.sort.insertion(EntityFreq, results.items, {}, struct {
        fn lessThan(ctx: void, a: EntityFreq, b: EntityFreq) bool {
            _ = ctx;
            return a.frequency > b.frequency;
        }
    }.lessThan);

    const slice = try results.toOwnedSlice(allocator);
    for (slice) |item| {
        _ = freq_map.remove(item.name);
    }
    return slice;
}

const testing = std.testing;

test "candidateEntityWords extracts capitalized words" {
    const text = "The GuidanceDb struct implements search functionality";
    const candidates = try candidateEntityWords(text, testing.allocator);
    defer {
        for (candidates) |c| testing.allocator.free(c);
        testing.allocator.free(candidates);
    }
    var found_guidance = false;
    for (candidates) |c| {
        if (std.mem.eql(u8, c, "GuidanceDb")) found_guidance = true;
    }
    try testing.expect(found_guidance);
}

test "candidateEntityWords filters stoplist" {
    const text = "The function returns This value";
    const candidates = try candidateEntityWords(text, testing.allocator);
    defer {
        for (candidates) |c| testing.allocator.free(c);
        testing.allocator.free(candidates);
    }
    var has_stop = false;
    for (candidates) |c| {
        if (std.mem.eql(u8, c, "The") or std.mem.eql(u8, c, "This")) has_stop = true;
    }
    try testing.expect(!has_stop);
}

test "extractEntities with min_frequency" {
    const text = "GuidanceDb handles search. GuidanceDb is fast. OtherClass is also here.";
    const entities = try extractEntities(text, testing.allocator, 2);
    defer {
        for (entities) |e| testing.allocator.free(e.name);
        testing.allocator.free(entities);
    }
    var found = false;
    for (entities) |e| {
        if (std.mem.eql(u8, e.name, "GuidanceDb")) found = true;
    }
    try testing.expect(found);
}

File: src/ontology/src/entity.rs
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EntityType {
    Person,
    Project,
    Location,
    Uncertain,
}

#[derive(Debug, Clone)]
pub struct EntityFreq {
    pub name: String,
    pub frequency: usize,
    pub entity_type: EntityType,
}

use std::sync::LazyLock;

static ENTITY_STOPLIST: LazyLock<std::collections::HashSet<&'static str>> = LazyLock::new(|| {
    let mut s = std::collections::HashSet::new();
    for w in &[
        "The",
        "This",
        "That",
        "When",
        "Where",
        "What",
        "Why",
        "Who",
        "Which",
        "How",
        "Then",
        "There",
        "Here",
        "Now",
        "Just",
        "Also",
        "Some",
        "Such",
        "Each",
        "Every",
        "Monday",
        "Tuesday",
        "Wednesday",
        "Thursday",
        "Friday",
        "Saturday",
        "Sunday",
        "January",
        "February",
        "March",
        "April",
        "May",
        "June",
        "July",
        "August",
        "September",
        "October",
        "November",
        "December",
    ] {
        s.insert(*w);
    }
    s
});

fn is_capitalized(s: &str) -> bool {
    s.chars().next().is_some_and(char::is_uppercase)
}

fn is_entity_word_char(c: char) -> bool {
    c.is_alphanumeric() || c == '_' || c == '-' || c == '\'' || c == '.'
}

pub fn candidate_entity_words(text: &str) -> Vec<String> {
    let mut words = Vec::new();
    let mut current = String::new();
    for c in text.chars() {
        if is_entity_word_char(c) {
            current.push(c);
        } else {
            if current.len() >= 2
                && is_capitalized(&current)
                && !ENTITY_STOPLIST.contains(current.as_str())
            {
                words.push(current.clone());
            }
            current.clear();
        }
    }
    if current.len() >= 2 && is_capitalized(&current) && !ENTITY_STOPLIST.contains(current.as_str())
    {
        words.push(current);
    }
    words
}

pub fn extract_entities(content: &str, min_frequency: usize) -> Vec<EntityFreq> {
    let mut freq: HashMap<String, usize> = HashMap::new();
    for word in candidate_entity_words(content) {
        *freq.entry(word).or_insert(0) += 1;
    }
    let mut result: Vec<EntityFreq> = freq
        .into_iter()
        .filter(|(_, count)| *count >= min_frequency)
        .map(|(name, frequency)| EntityFreq {
            name,
            frequency,
            entity_type: EntityType::Uncertain,
        })
        .collect();
    result.sort_by_key(|b| std::cmp::Reverse(b.frequency));
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extracts_capitalized_words() {
        let words = candidate_entity_words("Alice met Bob at Google");
        assert!(words.contains(&"Alice".to_string()));
        assert!(words.contains(&"Bob".to_string()));
        assert!(words.contains(&"Google".to_string()));
    }

    #[test]
    fn filters_stoplist() {
        let words = candidate_entity_words("The quick brown Fox");
        assert!(!words.contains(&"The".to_string()));
        assert!(words.contains(&"Fox".to_string()));
    }

    #[test]
    fn extract_entities_with_min_frequency() {
        let text = "Alice and Alice and Bob and Alice";
        let entities = extract_entities(text, 2);
        let alice = entities.iter().find(|e| e.name == "Alice").unwrap();
        assert_eq!(alice.frequency, 3);
    }
}
